spatialdata/_core/spatialdata.py

```diff
@@ -1,25 +1,242 @@
+
+    def write(self, file_path: str) -> None:
+        """Write to Zarr file."""
+
+        store = parse_url(file_path, mode="w").store
+        root = zarr.group(store=store)
+
+        # get union of unique ids of all elements
+        elems = set().union(*[set(i.keys()) for i in self.__dict__.values()])
+
+        for el in elems:
+            elem_group = root.create_group(name=el)
+            if self.images is not None and el in self.images.keys():
+                self.images[el].to_zarr(elem_group, name=el)
+            if self.labels is not None and el in self.labels.keys():
+                self.labels[el].to_zarr(elem_group, name=el)
+            if self.points is not None and el in self.points.keys():
+                self.points[el].to_zarr(elem_group, name=el)
+            if self.polygons is not None and el in self.polygons.keys():
+                self.polygons[el].to_zarr(elem_group, name=el)
+            # TODO: shall we write tables?
+            # if el in self.tables.keys():
+            #     self.tables[el].to_zarr(elem_group, name=el)
+
+    def __repr__(self) -> str:
+        return self._gen_repr()
+
+    def _gen_repr(
+        self,
+    ) -> str:
+        def rreplace(s: str, old: str, new: str, occurrence: int) -> str:
+            li = s.rsplit(old, occurrence)
+            return new.join(li)
+
+        def h(s: str) -> str:
+            return hashlib.md5(repr(s).encode()).hexdigest()
+
+        descr = "SpatialData object with:"
+        for attr in ["images", "labels", "points", "polygons", "tables"]:
+            attribute = getattr(self, attr)
+            descr += f"\n{h('level0')}{attr.capitalize()}"
+            descr = rreplace(descr, h("level0"), "└── ", 1)
+            if attribute is not None:
+                for k, v in attribute.items():
+                    descr += f"{h('empty_line')}"
+                    if isinstance(v, AnnData):
+                        descr_class = v.__class__.__name__
+                    else:
+                        descr_class = v.data.__class__.__name__
+                    descr += f"{h('level1.0')}'{k}': {descr_class} {v.shape}"
+                    descr = rreplace(descr, h("level1.0"), "    └── ", 1)
+            if attr == "tables":
+                descr = descr.replace(h("empty_line"), "\n  ")
+            else:
+                descr = descr.replace(h("empty_line"), "\n│ ")
+        return descr
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, SpatialData):
+            return False
+        # new comparison: dumping everything to zarr and comparing bytewise
+        # TODO: do the same but in memory
+        with tempfile.TemporaryDirectory() as tmpdir:
+            self.write(os.path.join(tmpdir, "self.zarr"))
+            other.write(os.path.join(tmpdir, "other.zarr"))
+            return are_directories_identical(os.path.join(tmpdir, "self.zarr"), os.path.join(tmpdir, "other.zarr"))
```

Comment:
I recognize this is a temporary solution, but I think I would rather not have an `__eq__` method than have one that writes objects to disk. 

In general, I think we want the `__eq__` methods that do the comparison to live on the elements and this `SpatialData` `__eq__` just call those.

Reply:
I agree, but comparing anndata objects is not possible, so at least for that we may need to compare files on disk.

Reply:
> comparing anndata objects is not possible

@LucaMarconato , can you please expand on this? Do you mean we would need to contribute an `AnnData.__eq__()` method upstream to anndata? Or is there some other limitation?

In any case, I could imagine it would be slow since there's potentially a lot of data to compare. Perhaps we should just remove this method?

Reply:
jumping in only on this comment: indeed anndata doesn't have such method, so we'd ahve to do it upstream (although I believe there is plans to do it @ivirshup ). Regardless, I agree we;d want to remove it eventually, but for now I think it can turns out quite handy for prototyping ?

Reply:
Overall, my concern here is that the `__eq__` method is part of the public API and I think this implementation is really not ideal (e.g., this will cause problems if a user calls this on reasonably large data). Even though we are early stage, my understanding is that we want to shop `SpatialData` around to potential users relatively soon. Thus, I think we need to be careful about what we put on our public API. 

> indeed anndata doesn't have such method, so we'd ahve to do it upstream (although I believe there is plans to do it @ivirshup ). Regardless, I agree we;d want to remove it eventually, but for now I think it can turns out quite handy for prototyping ?

Can you expand on what you would like to use this for? Could this be a private utility function? If anndata doesn't have an `__eq__` implementation, that is a fairly strong signal to me that we don't need one either (yet).

Reply:
I think if AnnData gets an `__eq__` method, it's probably not going to return a single boolean. It'd probably be broadcasted over all the elements.

For something that returns a bool, it'd be a separate function. I think it's likely we'll have this one first since we've already got `assert_equal` for testing. It'll probably be slow and kept in a testing module.

Reply:
Ok thanks for the comments, I'll turn it into a private method.

Reply:
Sounds good. Thanks everyone!

Reply:
I'll remove it here.